File: 81-search-in-rotated-sorted-array-ii/search-in-rotated-sorted-array-ii.go
func search(nums []int, target int) bool {
    l := 0
    r := len(nums) - 1
    for l <= r {
        mid := ((r-l) >> 1) + l

        if nums[mid] == target {
            return true
        }

        if nums[l] == nums[mid] {
            l++
            continue
        }

        if nums[mid] <= nums[r] {
            if nums[mid] < target && target <= nums[r] {
                l = mid + 1
            } else {
                r = mid - 1
            }
        } else {
            if nums[l] <= target && target < nums[mid] {
                r = mid - 1
            } else {
                l = mid + 1
            }
        }
    }

    return false
}

File: 81-search-in-rotated-sorted-array-ii/search-in-rotated-sorted-array-ii.cpp
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int l = 0, r = nums.size() - 1;

        while (l <= r) {
            int mid = ((r - l) >> 1) + l;

            if (nums[mid] == target) {
                return true;
            }

            if (nums[l] == nums[mid]) {
                l++;
                continue;
            }

            if (nums[mid] <= nums[r]) {
                if (nums[mid] < target && target <= nums[r]) {
                    l = mid + 1;
                } else {
                    r = mid - 1;
                }
            } else {
                if (nums[l] <= target && target < nums[mid]) {
                    r = mid - 1;
                } else {
                    l = mid + 1;
                }
            }
        }
        return false;
    }
};